designs/disruption-controls-drain-timeout.md

```diff
@@ -0,0 +1,78 @@
+# Disruption Controls - (in v1beta1+)
+
+## Motivation
+Users are requesting the ability to control how long karpenter will wait for a drain on an individual node to complete before forcefully evicting the remaining pods to decommission the node [#743](https://github.com/kubernetes-sigs/karpenter/issues/743). This supports two primary use cases.
+* Cluster admins who want to ensure that nodes are cycled after a given period of time, regardless of user-defined disruption controls (such as PDBs or preStop hooks) that might prevent eviction of a pod beyond the configured limit. This could be to satisfy security requirements or for convenience.
+* Cluster admins who want to allow users to protect long-running jobs from being interrupted by node disruption, up to a configured limit.
+
+This design piggybacks off the existing disruption-controls design, adding in a new `forceDrainAfter` field to the `v1beta/NodePool` `Disruption` spec. This field seems intuitively related to the other disruption controls that exist within the disruption block.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    forceDrainAfter: 24h || Never
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // ForceDrainAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable force draining of pods.
+    ForceDrainAfter *NillableDuration `json:"forceDrainAfter" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+The `forceDrainAfter` fields accepts a common duration string which defaults to a value of `Never`. Omitting the field results in the default value instructing the controller to wait indefinitely for pods to drain gracefully, maintaining the existing behavior.
+
+## Prior Art
+This is the field from CAPI's MachineDeployment spec which implements identical behavior.
+```
+$ kubectl explain machinedeployment.spec.template.spec.nodeDrainTimeout
+KIND:     MachineDeployment
+VERSION:  cluster.x-k8s.io/v1beta1
+
+FIELD:    nodeDrainTimeout <string>
+
+DESCRIPTION:
+     NodeDrainTimeout is the total amount of time that the controller will spend
+     on draining a node. The default value is 0, meaning that the node can be
+     drained without any time limitations. NOTE: NodeDrainTimeout is different
+     from `kubectl drain --timeout`
+```
+
+## Implementation
+
+### Force Drain Detection
+If forceDrainAfter is set, I think it should be straightforward to have the node termination controller add a timestamp as an annotation on the Node resource denoting when the Node will be force drained at a specific time (is RFC-3339 the preferred standard?)
```

Comment:
One more note is if we move the finalizer out if node then the above comment won't work https://github.com/kubernetes-sigs/karpenter/issues/758#issuecomment-1790931682 since deletion won't happen on the node first -> deletion timestamp won't exist on the node. However we likely won't just consult the nodeclaim if this migration happens

Reply:
I did want to avoid the use of an extra annotation for state tracking, but this is why I didn't think `deletionTimestamp` would work.

```
$ kubectl get node ip-10-117-32-4.us-east-2.compute.internal -o yaml | grep deletionTimestamp
<empty>

$ kubectl drain ip-10-117-32-4.us-east-2.compute.internal --delete-emptydir-data --ignore-daemonsets
node/ip-10-117-32-4.us-east-2.compute.internal already cordoned
...
<all pods cleanly evicted>
...
node/ip-10-117-32-4.us-east-2.compute.internal drained

$ kubectl get node ip-10-117-32-4.us-east-2.compute.internal -o yaml | grep deletionTimestamp
<empty>

<node is reaped by node autoscaler for being unschedulable and empty>
```
`deletionTimestamp` isn't set until the Node resource is deleted, which doesn't occur during a normal drain.

1. Does Karpenter always perform a Node deletion as part of its node deprovisioning process?
2. If so, does Karpenter perform the Node deletion at the start of the deprovisioning process, or it is only performed after all pods have been successfully drained from the node?

Reply:
https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/#understanding-finalizers about deletion with finalizers 
```
Kubernetes will report back that the object has been deleted, however, it hasn’t been deleted in a traditional sense. Rather, it’s in the process of deletion. When we attempt to get that object again, we discover the object has been modified to include the deletion timestamp.```

Reply:
I think the standard best practice for K8s node cleanup is to completely drain the Node before any deletion occurs.
https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/

I think that agrees with the comment you linked above?
> Minimally, we need to stop deleting the node and allow the kubelet to deregister as part of our standard termination workflow, as it's currently causing daemons to leak.

Reply:
Agreed on thats what we should do however that isn't what Karpenter is doing atm,  until the drain happens at nodeclaim first then node. I believe this isn's something that's settled on yet. So this depends on how we want to implement this to match the future of Karpenter. It would be good to know if Karpenter will end up migrating the finalizer. (I am ok with annotation for now).

Ideally I think it shouldn't be an annotation but rather a status on the nodeclaim.

Reply:
The finalizer should be both on the node and nodeclaim. Whichever receives the deletion request, either by the user or by Karpenter, should always trigger the deletion behavior. The [termination code](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/node/termination/controller.go#L75) requires this deletion timestamp to be set to know when a delete request has been made.  

I don't see a reason why we can't rely on the deletion timestamp here. If there's a debate on whether we should rely on the nodeclaim's or node's deletion timestamp, Karpenter deletes the nodeclaim to then delete the node, whereas a user could do either. Since both nodeclaim and node deletion trigger deletion of the other linked resource, I could imagine picking one of the two as the source of truth (which I would argue we should do NodeClaim, since it's the resource that Karpenter chooses to represent both the node and underlying machine).

Reply:
I think using the NodeClaim's `deletionTimestamp` should be sufficient.

Reply:
> The finalizer should be both on the node and nodeclaim

Having the finalizer on the NodeClaim gives us the benefit of removing the finalizer on the Node later. We've talked a little bit about potentially removing it since it's caused some node hanging in cases where the karpenter controller has been removed. I'm not fully convinced that we want to remove the node finalizer, but I do think there's quite a bit of value in pushing more detail into the NodeClaim management object over the Node itself